Deprecate Obj.set_tag

stedolan opened the pull request:
Under the multicore GC, object headers (tags and lengths) are immutable. Not in the current sense of "don't change these unless you know what you're doing", but in the stronger sense of "if you ever change these, the GC will randomly segfault as it scans the object you're messing with from another thread". Consequently, `Obj.set_tag` is unsupported. (`Obj.truncate` too, although this PR only talks about the former).

Currently, this PR doesn't contain any code. Before doing a bunch of hacking, I'd like to discuss with the maintainers (cc @xavierleroy, @damiendoligez) whether (a) deprecating and later removing `set_tag` is something they find acceptable in principle and (b) whether the below is a reasonable plan for doing it.

This change will break code that uses `Obj.set_tag`, although I'd argue that occasional breakage is the price of using `Obj`. Still, I did some grepping of the stdlib and the OPAM universe to find uses of `set_tag`, and work out replacements for common uses:

### Lazy

[`CamlinternalLazy` uses `set_tag`](https://github.com/ocaml/ocaml/blob/trunk/stdlib/camlinternalLazy.ml#L30) to change `Lazy_tag` to `Forward_tag` when thunks are forced.

I suggest using an extra word in the `Lazy.t` structure to distinguish forced and unforced thunks, which will increase the size of unforced thunks from 2 to 3 words. For thunks that are eventually forced the increase is temporary, since the GC short-circuits forced thunks during tracing.

### Creating new blocks with unexpected tags

Some code ([example](https://github.com/janestreet/typerep_extended/blob/master/src/type_struct.ml#L1461-L1462)) uses `set_tag` to create a block with a tag other than what OCaml normally chooses. ([@alainfrisch has also written](https://www.lexifi.com/blog/references-physical-equality) about using this trick to get polymorphic comparison to understand hashconsing, although I haven't come across uses in the wild).

For these uses, a new primitive `Obj.with_tag : int -> Obj.t -> Obj.t` suffices: like `dup`, but allows the user to specify an alternative tag for the copy. This still allows the construction of objects with unusual tags, but by creating a fresh object with the desired tag, rather than by mutation. If performance is a concern, an optimisation that detects when `with_tag` is applied directly to an allocation and elides the first allocation is easy enough to implement.

### Preventing the GC from scanning `int array`

At runtime, `int array` looks just like any other array, so gets needlessly scanned by the GC. Some programs ([example](https://github.com/c-cube/zipperposition/blob/master/src/core/Ordering.ml#L135)) explicitly change the tag of integer arrays to prevent scanning.

Unlike the other uses, this use of `set_tag` can be silently ignored without affecting correctness. It is a reasonable optimisation, though, and it would be nice to support it properly. Perhaps we could introduce a new `intarray` type (matching `floatarray`) which has the same layout as ordinary arrays, but is tagged with `No_scan_tag`. (Unlike `floatarray`, the layout is always compatible with standard arrays, so `= private int array` is a reasonable definition of this type).

### Unfixable cases

Some code is not so easily fixable, either by making optimisations based on undocumented properties of the current runtime ([example](https://github.com/OCamlPro/operf-micro/blob/master/share/operf-micro/benchmarks/nullable_array/nullable_array_closure.ml#L37-L44)), or by targeting not the OCaml language but a runtime model that explictly includes `set_tag` ([example](https://github.com/protz/mezzo/blob/master/compiler/UntypedMezzo2UntypedOCaml.ml#L225-L236)).

bluddy commented:
> Perhaps we could introduce a new intarray type (matching floatarray) which has the same layout as ordinary arrays, but is tagged with No_scan_tag.

There should be other cases like this. Perhaps it would be better to dedicate a bit in the 64-bit header to indicate whether the GC should scan the data.

damiendoligez commented:
> There should be other cases like this. Perhaps it would be better to dedicate a bit in the 64-bit header to indicate whether the GC should scan the data.

I don't think so. Whether data should be scanned by the GC or not, is highly dependent on the data itself, there is no orthogonality here.

alainfrisch commented:
> I suggest using an extra word in the Lazy.t structure to distinguish forced and unforced thunks, which will increase the size of unforced thunks from 2 to 3 words. For thunks that are eventually forced the increase is temporary, since the GC short-circuits forced thunks during tracing.

More generally, can you describe how lazy blocks are supported in the multicore GC?  Marking the block as a "forward" block and storing the forward pointer need to be done atomically from the point of view of other threads.  How do you achieve that?

The performance impact of the new fields should be evaluated.  The size increase could be noticeable for lazy-heavy data structures , and even if lazy thunks are forced quickly the extra allocations can put more pressure on the GC.  Also, scanning and other generic operations could become a bit slower.

> (Unlike floatarray, the layout is always compatible with standard arrays, so = private int array is a reasonable definition of this type).

I agree that adding intarray is a good idea, but I would refrain from exposing the fact that it's a subtype of "int array".  This would allow creating two int array values with the same elements, but which would be different when compared with the generic equality operator.

kayceesrk commented:
@alainfrisch Here is the issue that tracks Lazy proposal for Multicore OCaml https://github.com/ocamllabs/ocaml-multicore/issues/188

dra27 commented on stdlib/obj.mli:
```diff
@@ -55,6 +55,7 @@ external field : t -> int -> t = "%obj_field"
 *)
 external set_field : t -> int -> t -> unit = "%obj_set_field"
 external set_tag : t -> int -> unit = "caml_obj_set_tag"
+  [@@ocaml.deprecated "Allocated block tags are immutable, but see with_tag for new allocations."]
```
#1148 will cause this to be a CI error - please break the string up.

dra27 reviewed (dismissed):
I apologise hugely for requesting changes on a 1-line GPR!

stedolan committed:
Deprecate Obj.set_tag.

  - Adds Obj.with_tag as a partial replacement.
  - Adds caml_obj_make_forward for use of Camlinternal{Lazy,Mod}

stedolan committed:
Add optimisations in simplif.ml for Obj.with_tag.

stedolan commented:
New implementation just pushed.

This one keeps the current behaviour for `Lazy`, which was the most controversial part of the current design. Multicore's invariants have been weakened from "tags never change" to "tags never change unless they're `Lazy_tag`" (see ocamllabs/ocaml-multicore#226).

A new primitive, `Obj.with_tag` has been added to cover the use-cases of allocating an object with a nonstandard tag. There's also a new optimisation (with tests) in `simplif.ml` to ensure that patterns like `Obj.with_tag (Obj.repr { ... new allocation ... })` do not cause any more allocations than the original `set_tag`-based code.

gasche commented on bytecomp/simplif.ml:
```diff
@@ -235,6 +235,16 @@ let simplify_exits lam =
                                      ap_inlined=Default_inline;
                                      ap_specialised=Default_specialise}
+        (* Simplify %identity *)
+      | Pidentity, [e] -> e
```
Why is this line part of the present change?

stedolan replied:
The idiom being optimised is `Obj.with_tag (Obj.repr ( ... alloc ... ))` which has an extra `Pidentity` because `repr` expands to `%identity`. I thought that adding the general `Pidentity e -> e` transformation was cleaner than a composed `with_tag (Pidentity (Pmakeblock(...)) -> Pmakeblock(...)` transformation.

gasche commented on runtime/obj.c:
```diff
@@ -90,16 +97,16 @@ CAMLprim value caml_obj_block(value tag, value size)
 
   sz = Wosize_val(arg);
+  tg = (tag_t)Long_val(new_tag_v);
   if (sz == 0) CAMLreturn (arg);
```
I'm not sure that this `sz == 0` optimization is correct if the tag is different.

stedolan replied:
Wow, that's clearly wrong. Nice catch.

gasche commented on stdlib/camlinternalLazy.ml:
```diff
@@ -41,9 +41,7 @@ let force_val_lazy_block (blk : 'arg lazy_t) =
   let closure = (Obj.obj (Obj.field (Obj.repr blk) 0) : unit -> 'arg) in
   Obj.set_field (Obj.repr blk) 0 raise_undefined;
   let result = closure () in
-  (* do set_field BEFORE set_tag *)
```
What were these comments for? Signal safety? (Otherwise, should they be reproduced in the C implementation?)

stedolan replied:
The GC must not be allowed to observe `Forward_tag` before the field has been updated. This potentially mattered when the sequence was in OCaml (it's hard to predict when the GC will run), but the order does not matter in C (the sequence cannot invoke the GC, so nothing can observe the order).

gasche commented on stdlib/obj.mli:
```diff
@@ -55,6 +55,7 @@ external field : t -> int -> t = "%obj_field"
 *)
 external set_field : t -> int -> t -> unit = "%obj_set_field"
 external set_tag : t -> int -> unit = "caml_obj_set_tag"
+  [@@ocaml.deprecated "Use with_tag instead."]
```
It's not necessarily a great idea to deprecate a feature at the same version that the replacement is added: a maintainer that would follow the deprecation advice must force its users to upgrade to the latest version (or use conditionals). Could we wait for one version between `with_tag` and `set_tag`'s deprecation?

stedolan replied:
Happy to follow whatever deprecation policy works best here, but I don't really understand the issue. I thought `@@ocaml.deprecated` was the way to signal that a replacement is available? Is the concern here programs compiled with `-warn-error` breaking?

damiendoligez replied:
Normally, we would deprecate it now and remove it in the next deprecation purge, in about 5 years. We still don't have any mechanism to indicate degrees of deprecation (or how much time before it gets removed). Such a mechanism might be an intermediate step before deprecation.

In the case of `Obj.set_tag` I think we'll remove it out-of-cycle to avoid delaying multicore, so it's better to deprecate right now anyway.

dbuenzli replied:
> It's not necessarily a great idea to deprecate a feature at the same version that the replacement is added:

Regarding this comment about @gasche I fear a bit nothing is ever going to be deprecated (people will simply forget) if nothing special is done. One simple measure would be to have in the repo text file with a table of the form:

```
ID | vReplacement | vDeprecated notice | vRemove
--------------------------------------------------
```
Before a release one could then apply appropriate changes according to what the table says.

gasche replied:
@dbuenzli yes, that sounds like a good idea. I guess some of your merged PRs have some maybe-deprecate-later identifiers in flight, do you have enough to send a PR to populate that file? (Did we merge PrintExc successor yet?)

(The file should also point to the (M/G)PR that contains the deprecation action/discussion.)

nojb replied:
>  (Did we merge PrintExc successor yet?)

No, it is #2137, but it is not yet ready, more work needed.

dbuenzli replied:
I could have a look but a bit busy® right now®.

gasche commented:
I have an uneasy feeling when compiler code matches on the name of C functions. I suspect that the externals whose semantic is part of the language (enough that the compiler knows about them) should be `%foo` functions, that later get turned into C primitives by the unsupporting backends (possibly all backends) if desired.
However, I understand why the PR just uses a new C primitive, given that the rest of the Obj module is in this style (but then, it looks like there were no Obj-specific optimizations in the compiler before?).

stedolan commented:
I was also unsure about exposing this as `"caml_obj_with_tag"` vs. `"%obj_with_tag"`. In doubt, I went for the version that's less code, but I don't have a strong preference. Do you think this would be better as a primitive + later conversion to a C call?

stedolan committed:
Changes entry and review fixes

gasche commented:
Personally I think that all of Obj's primitives could/should be moved to `%obj_*` primitives instead of C calls, because we make assumptions on them, but that can be done separately. (And it may break compatibility of some libraries that re-export Obj, sigh...)

I also don't have very strong feelings on the best deprecation policy.

gasche commented on runtime/obj.c:
```diff
@@ -106,7 +106,7 @@ CAMLprim value caml_obj_with_tag(value new_tag_v, value arg)
   sz = Wosize_val(arg);
   tg = (tag_t)Long_val(new_tag_v);
-  if (sz == 0) CAMLreturn (arg);
+  if (sz == 0) CAMLreturn (Atom(tg));
```
(I wondered whether this change would degrade performance sensibly; `Atom(tg)` extracts the value from a precomputed table, so I don't think we would notice the difference, and it's not clear that we use `set_tag` on empty blocks that often.)

stedolan replied:
`Atom(tg)` doesn't actually touch the table: it returns the address of the table plus `tg` (0-field objects are pointers into the atom table). So I can't imagine this making any difference.

gasche reviewed (approved):
I reviewed the patch for correctness and I think the new primitive is a nice thing to have in any case. The only thing I'm unsure about is the idea of deprecating `set_tag` right away; for a more widely used API I would frown upon deprecated in the same version its replacement appears, but this might not make a big difference for such a low-level function?

I'm approving, but I think this deserves a third pair of eyes (and maybe a decision on deprecation policy) before merging.

stedolan committed:
check_typo fix

stedolan commented:
I don't understand the AppVeyor failure. It's complaining about the new C primitive, but the Travis build is fine. Perhaps a bootstrap would fix it? I don't understand why this differs on windows. @dra27 ?

nojb commented:
It fails when building `flexlink` (only used on Windows) using the bootstrap compiler. My understanding is that the only way out is a bootstrap.

stedolan committed:
Bootstrap

damiendoligez commented on testsuite/tests/lib-obj/with_tag.ml:
```diff
@@ -0,0 +1,31 @@
+  assert (Obj.with_tag 1 (Obj.repr (A ("hello", 10.))) = Obj.repr (B ("hello", 10.)))
+
+let () =
+  assert (Obj.tag (Obj.with_tag 42 (Obj.repr [| |])) = 42)
```
The empty array with tag 42 is going to be atom number 42, but (I think) atoms above 0 are currently not used by OCaml, so you are testing unused corners of the runtime. Maybe you would like to put an element in this array instead.

gasche replied:
I think it's the point of testing the empty atom, to avoid a regression that would have been present in an earlier version of the patch (where empty atoms were returned unchanged).

damiendoligez reviewed (approved):
Third pair of eyes here. Looks good except for a small (and optional) suggestion.
In this case I think it's OK to deprecate right away.

gasche commented:
@stedolan: you should add Damien as reviewer and then it's good to merge.

stedolan committed:
update Changes

gasche commented:
Merged, thanks!
